src/sentry_core.c

```diff
@@ -89,39 +90,61 @@ sentry_init(sentry_options_t *options)
     // enumeration.
     options->run = sentry__run_new(options->database_path);
     if (!options->run) {
-        sentry_options_free(options);
-        return 1;
+        SENTRY_WARN("failed to initialize run directory");
+        goto fail;
     }
 
     load_user_consent(options);
 
-    // we "activate" the options here, since the transport and backend might try
-    // to access them
+    if (transport) {
+        if (!sentry__transport_startup(transport, options)) {
+            SENTRY_WARN("failed to initialize transport");
+            goto fail;
+        }
+    }
+
+    // and then we will start the backend, since it requires a valid run
+    sentry_backend_t *backend = options->backend;
+    if (backend && backend->startup_func) {
+        SENTRY_TRACE("starting backend");
+        if (!backend->startup_func(backend, options)) {
+            SENTRY_WARN("failed to initialize backend");
+            goto fail;
+        }
+    }
+
     sentry__mutex_lock(&g_options_mutex);
     g_options = options;
     sentry__mutex_unlock(&g_options_mutex);
 
-    sentry_transport_t *transport = options->transport;
-    if (transport) {
-        sentry__transport_startup(transport, options);
+    // *after* setting the global options, trigger a scope and consent flush,
+    // since at least crashpad needs that.
+    // the only way to get a reference to the scope is by locking it, the macro
+    // does all that at once, including invoking the backends scope flush hook
+    SENTRY_WITH_SCOPE_MUT (scope) {
+        (void)scope;
+    }
+    if (backend && backend->user_consent_changed_func) {
+        backend->user_consent_changed_func(backend);
     }
 
     // after initializing the transport, we will submit all the unsent envelopes
     // and handle remaining sessions.
     sentry__process_old_runs(options);
```

Comment:
The old order of execution was: _options_ -> _transport_ -> _old runs_ -> _backend_.
The new order of execution is: _transport_ -> _backend_ -> _options_ -> _old runs_.

Are we sure that it is fine to clean up old runs after initializing the backend?

Reply:
The transport previously required the global options but that was fixed already. The problem with the backend is that it currently uses "public" API internally, which requires the global options.
The other option would be to set the options, and afterwards *unset* them again in case the backends fails to init, which would be observable from the outside, which is also weird.
The cleanest thing to do would be to explicitly pass the options around, but that is hard to do right now.

All-in-all, I think options should only be set once when the SDK was successfully initialized and is ready to be used.